cmd/query/app/apiv3/grpc_handler.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apiv3
+
+import (
+	"context"
+
+	"github.com/gogo/protobuf/types"
+
+	"github.com/jaegertracing/jaeger/cmd/query/app/querysvc"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v3"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// Handler implements api_v3.QueryServiceServer
+type Handler struct {
+	QueryService *querysvc.QueryService
+}
+
+var _ api_v3.QueryServiceServer = (*Handler)(nil)
+
+// GetTrace implements api_v3.QueryServiceServer's GetTrace
+func (h *Handler) GetTrace(request *api_v3.GetTraceRequest, stream api_v3.QueryService_GetTraceServer) error {
+	traceID, err := model.TraceIDFromString(request.GetTraceId())
+	if err != nil {
+		return err
+	}
+
+	trace, err := h.QueryService.GetTrace(stream.Context(), traceID)
+	if err != nil {
+		return err
+	}
+	resourceSpans := jaegerSpansToOTLP(trace.GetSpans())
+	return stream.Send(&api_v3.SpansResponseChunk{
+		ResourceSpans: resourceSpans,
+	})
+}
+
+// FindTraces implements api_v3.QueryServiceServer's FindTraces
+func (h *Handler) FindTraces(request *api_v3.FindTracesRequest, stream api_v3.QueryService_FindTracesServer) error {
+	query := request.GetQuery()
+	queryParams := &spanstore.TraceQueryParameters{
+		ServiceName:   query.GetServiceName(),
+		OperationName: query.GetOperationName(),
+		Tags:          query.GetAttributes(),
+		NumTraces:     int(query.GetNumTraces()),
+	}
+	if query.GetStartTimeMin() != nil {
+		startTimeMin, err := types.TimestampFromProto(query.GetStartTimeMin())
+		if err != nil {
+			return err
+		}
+		queryParams.StartTimeMin = startTimeMin
+	}
+	if query.GetStartTimeMax() != nil {
+		startTimeMax, err := types.TimestampFromProto(query.GetStartTimeMax())
+		if err != nil {
+			return err
+		}
+		queryParams.StartTimeMax = startTimeMax
+	}
+	if query.GetDurationMin() != nil {
+		durationMin, err := types.DurationFromProto(query.GetDurationMin())
+		if err != nil {
+			return err
+		}
+		queryParams.DurationMin = durationMin
+	}
+	if query.GetStartTimeMax() != nil {
```

Comment:
> Maybe could assert this in the test's mock reader.

I am not sure what you mean

Reply:
Because you had `query.GetStartTimeMax()` here instead of `query.GetDurationMax()`, could it be caught if the mock reader asserted on the actual `*spanstore.TraceQueryParameters` instance rather than `mock.AnythingOfType("*spanstore.TraceQueryParameters")`.

Reply:
probably not worth the effort, it was just a thought :)

Reply:
I see :). I have fixed the parameter.